staticdata: Refactor sysimage loading

topolarity opened the pull request:
Introduce `jl_image_buf_t` to represent the in-memory details of an unparsed system image and clean-up several global variables and split loading logic in staticdata.c so that we get (almost) everything we need from the sysimage handle at once.

This allows sysimage loading to be separated into three phases:
  1. Lookup the raw sysimage buffer as a `jl_image_buf_t`
  2. For .so sysimages, parse the sysimage and initialize JIT targets, etc.
  3. Finally load the sysimage into a `jl_image_t`

Care was taken to preserve the existing behavior of calling `jl_set_sysimg_so` to configure the sysimage before initializing Julia, although this is likely to be next on the chopping block in a follow-up.

Dependent on #57523.

topolarity committed:
staticdata: Refactor sysimage loading

Introduce `jl_image_buf_t` to represent the in-memory details of an unparsed
system image and clean-up several global variables and split loading logic
in staticdata.c so that we (almost) get everything we need from the sysimage
handle at once.

This allows sysimage loading to be separated into three phases:
  1. Lookup the raw sysimage buffer as a `jl_image_buf_t`
  2. For .so sysimages, parse the sysimage and initialize JIT targets, etc.
  3. Finally load the sysimage into a `jl_image_t`

Care was taken to preserve the existing behavior of calling `jl_set_sysimg_so`
to configure the sysimage before initializing Julia, although this is likely
to be next on the chopping block in a follow-up.

vtjnash reviewed (approved):
I like this feels less magic and global coupling of hidden logic. Much clearer now

topolarity committed:
Pass around `jl_options.cpu_target` explicitly

This (mostly) removes the only usage of `jl_options` internally in
`processor_*.cpp`
